### PR TITLE
[flang] Retain internal and BIND(C) host procedure link in FIR

### PR DESCRIPTION
**Pull request by jeanPerier**
Currently, it is not possible to find back which fun.func is the host procedure of some internal procedure because the mangling of the internal procedure does not contain info about the BIND(C) name of the host.
This info may be useful to ensure dwarf DW_TAG_subprogram of internal procedures are nested under DW_TAG_subprogram  of host procedures for instance.

### CODE REVIEW

File: flang/lib/Lower/CallInterface.cpp [hunk 1]
```diff
@@ -589,6 +589,21 @@ void Fortran::lower::CalleeInterface::setFuncAttrs(
 static void addSymbolAttribute(mlir::func::FuncOp func,
                                const Fortran::semantics::Symbol &sym,
                                mlir::MLIRContext &mlirContext) {
+  // The link between an internal procedure and its host procedure is lost
+  // in FIR if the host is BIND(C) since the internal mangling will not
+  // allow retrieving the host bind(C) name, and therefore func.func symbol.
+  // Preserve it as an attribute so that this can be later retrieved.
+  if (sym.owner().kind() == Fortran::semantics::Scope::Kind::Subprogram)
+    if (const Fortran::semantics::Symbol *hostProcedure = sym.owner().symbol())
+      if (Fortran::semantics::IsBindCProcedure(*hostProcedure)) {
```
Comment (vzakhari) [inline flang/lib/Lower/CallInterface.cpp:598]: Just a suggestion: maybe we should just always set it for consistency.  I think this should not affect readability of FIR too much.

Comment (jeanPerier) [reply flang/lib/Lower/CallInterface.cpp]: Thanks for the feedback. Makes sense to me. It is also then possible to get rid of fir.internal_proc and just use the new attribute. I updated the patch to do it for all internal procedures, including the ones from the main programs, and removed fir.internal_proc.

**Review by vzakhari — APPROVED** [review]
Thank you, Jean! Sorry for the delay, it went under the radar.

**Comment by jeanPerier** [comment]
> Thank you, Jean! Sorry for the delay, it went under the radar.

No problem, I am the one who took ages to update it, thanks for the review!

**Comment by DominikAdamski** [comment]
Hi,
Your PR breaks compilation of OpenMP Fortran code:
``` 
program test
use, intrinsic :: iso_fortran_env
implicit none
    integer, parameter :: n = 1024
    integer(int32) :: x(n),  y(n)
    integer :: i
    x = 1
    y = 0
    y = test_offload(x)
    do i = 1,n
      if ( x(i) .ne. y(i)) then
        stop 1
      endif
    enddo
    print *, "Test passed"

    contains
        function test_offload(xin) result(xout)
            integer(int32), intent(in) :: xin(1024)
            integer(int32) :: xout(1024)
            integer :: ii
            xout = 0.0
!$omp target parallel do map(to:xin) map(from: xout)
            do ii=1,1024
                xout(ii) = xin(ii)
            end do
!$omp end target parallel do
        end function test_offload
end program test
```
Compilation command:
`flang-new   -O2  -fopenmp --offload-arch=sm_70  test.f95 -o main`

I observe ICE when flang-new tries to lower MLIR code to LLVM IR.

**Comment by jeanPerier** [comment]
> Hi, Your PR breaks compilation of OpenMP Fortran code:
> I observe ICE when flang-new tries to lower MLIR code to LLVM IR.

Thanks, I could reproduce. The OMPFunctionFiltering assumed all function symbol uses are calls/address_of. When processing the host program (in the Fortran sense, not OpenMP) it now erases its internal procedures as if it were a call, and then proceeds to process the erased internal function, hence the segfault.

Working on a fix.

**Pull request merged**